### PR TITLE
Use `monkeypatch.setenv` when mocking environment variables (2)

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Use `monkeypatch.setenv` when mocking environment variables (2).

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
<details><summary>Details</summary>
<p>

```python
"""
python refact.py $(git ls-files tests/**/*.py)
"""
from __future__ import annotations

import random
import ast
import openai
import asyncio
import re
import textwrap
import click
import logging
from typing import Any, Iterator, TypeVar, Dict
from functools import total_ordering

T = TypeVar("T")

logger = logging.getLogger(__name__)


def get_indent(s: str) -> int:
    return len(s) - len(s.lstrip())


class Finder(ast.NodeVisitor):
    """
    Finds function nodes that contain a mock.patch.dict call
    """

    def __init__(self, filename):
        self.filename = filename
        self.funcs = []
        self.bad_funcs = []

    def node_to_location(self, node) -> str:
        return f"{self.filename}:{node.lineno}:{node.col_offset}"

    def visit_FunctionDef(self, node) -> None:
        self.funcs.append(node)
        super().generic_visit(node)
        self.funcs.pop()

    @property
    def current_func(self) -> ast.FunctionDef | None:
        if self.funcs:
            return self.funcs[-1]
        return None

    def visit_With(self, node) -> None:
        for item in node.items:
            call = item.context_expr
            if (
                isinstance(call, ast.Call)
                and isinstance(call.func, ast.Attribute)
                and call.func.attr == "dict"
                and isinstance(call.func.value, ast.Attribute)
                and call.func.value.attr == "patch"
                and isinstance(call.func.value.value, ast.Name)
                and call.func.value.value.id == "mock"
            ):
                for arg in call.args:
                    # Find os.environ
                    if (
                        isinstance(arg, ast.Attribute)
                        and arg.attr == "environ"
                        and isinstance(arg.value, ast.Name)
                        and arg.value.id == "os"
                    ):
                        print(self.node_to_location(self.funcs[-1]))
                        if self.current_func and self.current_func not in self.bad_funcs:
                            self.bad_funcs.append(self.funcs[-1])
                        break
        self.generic_visit(node)


def join_lines(lines: list[str]) -> str:
    return "\n".join(lines)


@total_ordering
class Range:
    def __init__(self, start: int, end: int) -> None:
        self.start = start
        self.end = end

    def offset(self, offset: int) -> Range:
        return Range(self.start + offset, self.end + offset)

    def __lt__(self, other: Range) -> bool:
        return self.start < other.start

    @classmethod
    def from_node(cls, node: ast.AST) -> Range:
        return cls(node.lineno - 1, node.end_lineno - 1)

    def slice(self, lines: list[str]) -> list[str]:
        return lines[self.start : self.end + 1]


class Patch:
    def __init__(self, rng: Range, lines: list[str]) -> None:
        self.range = rng
        self.lines = lines

    def diff(self) -> int:
        """
        How many lines does this patch add or remove if applied to a file?
        """
        return len(self.lines) - (self.range.end - self.range.start + 1)

    def offset(self, offset: int) -> Patch:
        return Patch(self.range.offset(offset), self.lines)

    def __repr__(self) -> str:
        c = join_lines(self.lines)
        return f"""\
lineno: {self.range.start}
end_lineno: {self.range.end}
-----
{c}
-----
"""

    def apply(self, lines: list[str]) -> list[str]:
        return [
            *lines[: self.range.start],
            *self.lines,
            *lines[(self.range.end + 1) :],
        ]

    @staticmethod
    def batch_apply(lines: list[str], patches: list[Patch]) -> list[str]:
        offset = 0
        for patch in patches:
            p = patch.offset(offset)
            lines = p.apply(lines)
            offset += patch.diff()
        return lines


def is_valid_code(code: str) -> bool:
    """
    Is `code` valid Python code?
    """
    try:
        ast.parse(textwrap.dedent(code))
        return True
    except SyntaxError:
        return False


class TopLevelImportFinder(ast.NodeVisitor):
    """
    Finds top-level import statements
    ```
    import a  # top-level
    import b  # top-level

    def foo():
        import c  # not top-level
    ```
    """

    def __init__(self):
        self.imports = []

    def visit_Import(self, node: ast.Import) -> None:
        self.imports.append(node)

    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
        self.imports.append(node)

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        pass


def find_imports(code: str) -> list[int]:
    finder = TopLevelImportFinder()
    finder.visit(ast.parse(textwrap.dedent(code)))
    line_nums = set()
    for imp in finder.imports:
        line_nums.update(range(imp.lineno, imp.end_lineno + 1))
    return sorted(line_nums)


def remove_imports(code: str) -> str:
    import_lines = find_imports(code)
    return join_lines(l for i, l in enumerate(code.split("\n")) if i + 1 not in import_lines)


def extract_code_from_block(s: str) -> str | None:
    if m := re.search(r"```(?:python)?\n(.+)\n```", s, re.DOTALL):
        return m.group(1)
    return None


async def chat_complete(func_lines: list[str]) -> str:
    prompt = """
The following pytest function contains `with mock.patch.dict(os.environ, dictionary, ...):`. We can use
`monkeypatch.setenv` or `monkepatch.delenv` instead. Please rewrite the function using `monkeypatch.setenv` or
`monkeypatch.delenv`. If the function arguments doesn't contain `monkeypatch`, please add it.
You might be tempted to add missing import statements. Please don't do that.

```python
{code}
```

The output should look like this:

```python
<new_code>
```
""".format(
        code=join_lines(func_lines)
    )
    resp = await openai.ChatCompletion.acreate(
        model="gpt-3.5-turbo-0613",
        messages=[
            {"role": "user", "content": prompt},
        ],
        temperature=0.0,
    )
    return resp["choices"][0]["message"]["content"]


def extract_code(chat_resp: str) -> list[str] | None:
    logger.debug("Response: \n---\n%s\n---", chat_resp)
    new_code = extract_code_from_block(chat_resp)
    if new_code is None:
        return None
    if not is_valid_code(new_code):
        return None
    return new_code


def clean_up_code(code: str, indent: int) -> str:
    # When a function contains undefined modules, ChatGPT sometimes imports them even when
    # we don't ask it to. Those imports need to be removed.
    return remove_imports(textwrap.indent(code, " " * indent))


def batched(iterable: Iterator[T], n: int) -> Iterator[list[T]]:
    batch: list[T] = []
    for item in iterable:
        batch.append(item)
        if len(batch) == n:
            yield batch
            batch = []
    if batch:
        yield batch


def collect_funcs(files: list[str]) -> Dict[str, list[Range]]:
    funcs_by_files = {}
    for file in files:
        with open(file) as f:
            code = f.read()

        tree = ast.parse(code)
        visitor = Finder(file)
        visitor.visit(tree)

        for func in visitor.bad_funcs:
            funcs_by_files.setdefault(file, []).append(Range.from_node(func))
    return funcs_by_files


async def fix(files: list[str]) -> None:
    print("Collecting functions to fix")
    tasks_by_file = collect_funcs(files)
    print("=" * 30)
    print("Generating patches")
    for file, ranges in tasks_by_file.items():
        with open(file) as f:
            lines = f.read().split("\n")

        patches: list[Patch] = []
        for batch in batched(ranges, 10):
            for _ in range(3):  # Retry 3 times
                coros = (chat_complete(rng.slice(lines)) for rng in batch)
                results = await asyncio.gather(*coros, return_exceptions=True)
                failed: list[Range] = []
                for rng, resp in zip(batch, results):
                    if isinstance(resp, Exception):
                        failed.append(rng)
                        logger.error("Failed to generate patch for %s", rng)
                        continue

                    if code := extract_code(resp):
                        indent = get_indent(rng.slice(lines)[0])
                        code = textwrap.indent(code, " " * indent)
                        code = remove_imports(code)
                        patches.append(Patch(rng, code.split("\n")))

                if not failed:
                    break
                batch = failed

        patches.sort(key=lambda p: p.range)
        print(patches)
        new_lines = Patch.batch_apply(lines, patches)
        with open(file, "w") as f:
            f.write(join_lines(new_lines))


@click.command()
@click.option("--debug", is_flag=True, default=False, help="Turn on debug logging")
@click.argument("files", nargs=-1)
def main(debug: bool, files: tuple[str]) -> None:
    if debug:
        logging.basicConfig(level=logging.DEBUG)
    files = list(files)
    random.shuffle(files)
    asyncio.run(fix(files))


if __name__ == "__main__":
    main()

```

</p>
</details>

File: tests/store/artifact/test_dbfs_artifact_repo_delegation.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def host_creds_mock():
 
 @mock.patch("mlflow.utils.databricks_utils.is_dbfs_fuse_available")
 def test_dbfs_artifact_repo_delegates_to_correct_repo(
-    is_dbfs_fuse_available, host_creds_mock
+    monkeypatch, is_dbfs_fuse_available, host_creds_mock
```
Comment (harupy) [inline tests/store/artifact/test_dbfs_artifact_repo_delegation.py:24]:
```suggestion
    is_dbfs_fuse_available, monkeypatch, host_creds_mock
```

File: tests/server/test_handlers.py [hunk 2]
```diff
@@ -227,21 +226,18 @@ def test_handler():
...
-    env = {
-        BACKEND_STORE_URI_ENV_VAR: f"file-plugin:{tmp_path}",
-    }
-    with mock.patch.dict(os.environ, env):
+    monkeypatch.setenv(BACKEND_STORE_URI_ENV_VAR, f"file-plugin:{tmp_path}")
+    mlflow.server.handlers._tracking_store = None
+    try:
+        plugin_file_store = mlflow.server.handlers._get_tracking_store()
+    finally:
         mlflow.server.handlers._tracking_store = None
```
Comment (harupy) [inline tests/server/test_handlers.py:238]: /ai rewrite this code using `monkeypatch.setattr`

Comment (harupy) [reply tests/server/test_handlers.py:238]:
```suggestion
    monkeypatch.setenv(BACKEND_STORE_URI_ENV_VAR, f"file-plugin:{tmp_path}")
    monkeypatch.setattr(mlflow.server.handlers, "_tracking_store", None)
```

**Review by WeichenXu123 — APPROVED** [review]
LGTM

**Pull request merged**